packages/amplify-category-function/src/provider-utils/awscloudformation/utils/truncateResourceNames.ts

```diff
@@ -0,0 +1,53 @@
+import { v5 as uuid } from 'uuid';
+
+const funcNameMaxLen = 64; // max length allowed by Lambda service
+const roleNameMaxLen = 64; // max length allowed by IAM service
+
+type Params = {
+  functionName?: string;
+  roleName?: string;
+};
+
+export const truncateResourceNames = (params: Readonly<Params>): Params => {
+  const result: Params = {};
+  if (typeof params.functionName === 'string') {
+    result.functionName = ResourceNameTruncator.withLimit(funcNameMaxLen).truncate(params.functionName);
+  }
+
+  if (typeof params.roleName === 'string') {
+    result.roleName = ResourceNameTruncator.withLimit(roleNameMaxLen).truncate(params.roleName);
+  }
+  return result;
+};
+
+class ResourceNameTruncator {
+  // this is an arbitrary uuid that is used as a seed to generate deterministic hashes from a given resource name
+  // eslint-disable-next-line spellcheck/spell-checker
+  private readonly uuidSeed = '319569b2-7cdc-4712-8390-e22b1f6ce5a9';
```

Comment:
nit: might be just my preference, have you considered passing the seed as argument to truncate method (or is there other reason to have it deterministic here?), this way you could hard-code it in tests but pass dynamic at runtime?

Reply:
the seed needs to be constant at runtime as well because we need to generate the same truncated name for the same long name every time. Otherwise, we would be tearing down and spinning up new lambdas on every push even if nothing changed